ports/esp8266/modneopixel.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Hubert Denkmair
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/mpconfig.h"
+#if MICROPY_ESP8266_NEOPIXEL
+
+#include "py/runtime.h"
+#include "esp_mphal.h"
+#include "espneopixel.h"
+
+typedef struct {
+    mp_obj_base_t base;
+    pyb_pin_obj_t* pin;
+    unsigned n;
+    uint8_t bpp;
+    size_t buf_size;
+    uint8_t *buf;
+} NeoPixel_obj_t;
+
+STATIC const mp_obj_type_t NeoPixel_type;
+STATIC mp_obj_t NeoPixel_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 2, 3, true);
+
+    NeoPixel_obj_t *self = m_new_obj(NeoPixel_obj_t);
+    self->base.type = &NeoPixel_type;
+    self->pin = mp_obj_get_pin_obj(args[0]);
+    self->n = mp_obj_get_int(args[1]);
+    self->bpp = (n_args > 2) ? mp_obj_get_int(args[2]) : 3;
+    self->buf_size = self->bpp * self->n;
+    self->buf = (uint8_t*)m_new(uint8_t, self->buf_size);
+    mp_hal_pin_output(self->pin->phys_port);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC void NeoPixel_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    (void)kind;
+    NeoPixel_obj_t* self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "<NeoPixel %p>", self->buf);
+}
+
+STATIC void NeoPixel_assert_color(NeoPixel_obj_t* self, size_t tuple_len) {
+    if (tuple_len != self->bpp) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "expecting a tuple of len bpp=%d", tuple_len));
+    }
+}
+
+STATIC mp_obj_t NeoPixel_fill(mp_obj_t self_in, mp_obj_t val_in) {
+    NeoPixel_obj_t* self = MP_OBJ_TO_PTR(self_in);
+
+    size_t val_len;
+    mp_obj_t *val;
+    mp_obj_get_array(val_in, &val_len, &val);
+    NeoPixel_assert_color(self, val_len);
```

Comment:
If you use `mp_obj_get_array_fixed_n(val_in, self->bpp, &val);` then it will automatically check the size of the tuple and there will be no need for the `NeoPixel_assert_color` function.